_example/envconfig.go

```diff
@@ -15,7 +17,9 @@ type ExampleEnvConfig struct {
 func (e *ExampleEnvConfig) Validate(_ diag.Component) diag.Diagnostics {
 	var diags diag.Diagnostics
 	if e.Port < 0 || e.Port > 65535 {
-		diags.Env("PORT").Error("Unable to parse port", "Port must be between 0 and 65535")
+		diags.Env("PORT").Error("Unable to parse port",
+			"Port must be between 0 and 65535, but instead got "+strconv.Itoa(e.Port))
+		e.Port = 0
```

Comment:
having to manually reset invalid parsed values to their empty values to avoid merging in invalid values is gross.

Reply:
Had a look around a different solutions, the way to go appears to be reflection.

```go
// Zero provides a utility which zeroes out the value of any standard type given
// a pointer.
func Zero(in any) {
	p := reflect.ValueOf(in).Elem()
	p.Set(reflect.Zero(p.Type()))
}
```

The problem is that you still have to manually check, making the code more verbose and much less performant due to using reflection:

```diff
if e.Port < 0 || e.Port > 65535 {
	diags.Env("PORT").Error("Unable to parse port",
		"Port must be between 0 and 65535, but instead got "+strconv.Itoa(e.Port))
-	e.Port = 0
+	configurator.Zero(&e.Port)
}
```

The only other option would be to move to using struct tags with validation rules, using something like https://github.com/go-playground/validator.

With this we could automate the validation and maybe the diag logging for the user. Meaning we could automate zeroing out the value if struct not valid. Again, this would use reflection, but if we can do it in a way that doesn't have to involve user intervention - that would be a win.

Outside the scope of this PR though.

Given that configurator makes no assumptions on how you do validation in the Validation method this enables the user to implement validator themselves.